.gitlab/includes/nvhpc_mpi_pipeline.yml

```diff
@@ -0,0 +1,64 @@
+# Copyright (c) 2023 ETH Zurich
+#
+# SPDX-License-Identifier: BSL-1.0
+# Distributed under the Boost Software License, Version 1.0. (See accompanying
+# file BOOST_LICENSE_1_0.rst or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+include:
+  - local: '.gitlab/includes/common_pipeline.yml'
+  - local: '.gitlab/includes/common_spack_pipeline.yml'
+
+.variables_nvhpc_mpi_config:
+  variables:
+    ARCH: linux-ubuntu22.04-zen3
+    BUILD_TYPE: Release
+    COMPILER: gcc@12.1.0
+    NVHPC_COMPILER: nvhpc@23.9
+    CXXSTD: 20
+    GPU_TARGET: '80'
+    SPACK_SPEC: "pika@main arch=$ARCH %${NVHPC_COMPILER} +stdexec +mpi +cuda cuda_arch=${GPU_TARGET} \
+                 malloc=system cxxstd=$CXXSTD ^boost@1.78.0 ^hwloc@2.7.0 \
+                 ^stdexec@git.nvhpc-23.09.rc4=main ^cuda@11.8 ^ninja%${COMPILER}"
+    CMAKE_FLAGS: "-DCMAKE_BUILD_TYPE=$BUILD_TYPE -DPIKA_WITH_CXX_STANDARD=$CXXSTD \
+                  -DPIKA_WITH_CUDA=ON -DPIKA_WITH_MALLOC=system -DPIKA_WITH_MPI=ON \
+                  -DCMAKE_CUDA_ARCHITECTURES=$GPU_TARGET -DPIKA_WITH_STDEXEC=ON"
+
+nvhpc_mpi_spack_compiler_image:
+  stage: spack_compiler
+  needs: [base_spack_image]
+  extends:
+    - .container-builder
+    - .variables_nvhpc_mpi_config
+  before_script:
+    - export DOCKERFILE_SHA=`sha256sum $DOCKERFILE | head -c 16`
+    - CONFIG_TAG=`echo $DOCKERFILE_SHA-$ARCH-$BASE_IMAGE-$COMPILER | sha256sum - | head -c 16`
+    - compiler=$COMPILER-$NVHPC_COMPILER
+    - compiler=${compiler//@/-}
+    - export PERSIST_IMAGE_NAME=$CSCS_REGISTRY_PATH/base/pika-$compiler:$CONFIG_TAG
+    - echo -e "compiler=$compiler" >> compiler.env
+    - echo -e "BASE_IMAGE=$PERSIST_IMAGE_NAME" >> compiler.env
+  variables:
+    DOCKERFILE: .gitlab/docker/Dockerfile.spack_compiler
+    DOCKER_BUILD_ARGS: '["BASE_IMAGE","ARCH","COMPILER","NVHPC_COMPILER"]'
+  artifacts:
+    reports:
+      dotenv: compiler.env
```

Comment:
Can this use the new templates instead?

Reply:
Good point. Maybe something like this:
```
~> COMPILER="gcc@11.4.0"; echo "${COMPILER}${NVHPC_COMPILER:+-}${NVHPC_COMPILER}"
gcc@11.4.0
~> COMPILER="gcc@11.4.0" NVHPC_COMPILER="nvhpc@23.9"; echo "${COMPILER}${NVHPC_COMPILER:+-}${NVHPC_COMPILER}"
gcc@11.4.0-nvhpc@23.9
```
? `${NVHPC_COMPILER:+-}` gets expanded to `-` if `NVHPC_COMPILER` is set (non-empty) and to nothing otherwise (https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html). I don't know how this interacts with `set -u` (not sure if it's set in that script) so you may need to experiment a bit, but I think it should be possible to get that to work.

Note by the way that you don't have `NVHPC_COMPILER` in the `CONFIG_TAG`. Maybe it would be good to add it there as well (to be consistent with everything else we put in `CONFIG_TAG`).